core/src/test/scala/unit/kafka/server/DeleteRecordsRequestTest.scala

```diff
@@ -0,0 +1,169 @@
+
+    // call the API
+    val response = sendDeleteRecordsRequest(request, leaderId)
+    val partitionResult = response.data.topics.find(invalidTopicPartition.topic).partitions.find(invalidTopicPartition.partition)
+
+    // Validate the expected error code in the response
+    assertEquals(Errors.UNKNOWN_TOPIC_OR_PARTITION.code(), partitionResult.errorCode(),
+      s"Unexpected error code received: ${Errors.forCode(partitionResult.errorCode()).name()}")
+  }
+
+  private def createTopicAndSendRecords = {
+    // Single topic
+    val topic1 = "topic-1"
+    val topicPartition = new TopicPartition(topic1, 0)
+    val partitionToLeader = createTopic(topic1)
+    assertTrue(partitionToLeader.contains(topicPartition.partition), "Topic creation did not succeed")
+    // Write records
+    produceData(Seq(topicPartition), MESSAGES_PRODUCED_PER_PARTITION)
+    (topicPartition, partitionToLeader(topicPartition.partition))
+  }
+
+  private def createDeleteRecordsRequestForTopicPartition(topicPartition: TopicPartition, offsetToDelete: Int) = {
+    val requestData = new DeleteRecordsRequestData()
+      .setTopics(Collections.singletonList(new DeleteRecordsTopic()
+        .setName(topicPartition.topic())
+        .setPartitions(Collections.singletonList(new DeleteRecordsPartition()
+          .setOffset(offsetToDelete)
+          .setPartitionIndex(topicPartition.partition())))))
+      .setTimeoutMs(TIMEOUT_MS)
+    val request = new DeleteRecordsRequest.Builder(requestData).build()
+    request
+  }
+
+  private def sendDeleteRecordsRequest(request: DeleteRecordsRequest, leaderId: Int): DeleteRecordsResponse = {
+    connectAndReceive[DeleteRecordsResponse](request, destination = brokerSocketServer(leaderId))
+  }
+
+  private def produceData(topicPartitions: Iterable[TopicPartition], numMessagesPerPartition: Int): Seq[RecordMetadata] = {
+    val records = for {
+      tp <- topicPartitions.toSeq
+      messageIndex <- 0 until numMessagesPerPartition
+    } yield {
+      val suffix = s"$tp-$messageIndex"
+      new ProducerRecord(tp.topic, tp.partition, s"key $suffix", s"value $suffix")
+    }
+    records.map(producer.send(_).get)
+  }
+
+  private def validateRecordsAreDeleted(topicPartition: TopicPartition, leaderId: Int, expectedOffset: Long): Unit = {
+    val targetTimes = List(new ListOffsetsTopic()
+      .setName(topicPartition.topic)
+      .setPartitions(List(new ListOffsetsPartition()
+        .setPartitionIndex(topicPartition.partition)
+        .setTimestamp(0L)).asJava)).asJava
+
+    val request = ListOffsetsRequest.Builder
+      .forConsumer(false, IsolationLevel.READ_UNCOMMITTED, false)
+      .setTargetTimes(targetTimes).build()
+
+    val listOffsetsPartitionResponse = connectAndReceive[ListOffsetsResponse](request, destination = brokerSocketServer(leaderId)).topics.asScala.find(_.name == topicPartition.topic).get
```

Comment:
Do you think we directly get offset from ReplicaManager instead of sending RPC?  for example, servers.map(_.replicaManager.getOffset).headOption

Reply:
Thank you for your review. These are good suggestions. I have removed the RPC and also added the tests for raft mode.

Reply:
Good suggestion! @dengziming !